launch_frontend/test/launch_frontend/test_substitutions.py

```diff
@@ -0,0 +1,136 @@
+        raise RuntimeError()
+    return TextSubstitution(text=''.join([i.perform(None) for i in data[0]]))
+
+
+def test_text_with_embedded_substitutions():
+    subst = default_parse_substitution('why_$(test asd)_asdasd_$(test bsd)')
+    assert len(subst) == 4
+    assert subst[0].perform(None) == 'why_'
+    assert subst[1].perform(None) == 'asd'
+    assert subst[2].perform(None) == '_asdasd_'
+    assert subst[3].perform(None) == 'bsd'
+
+# TODO(ivanpauno): Don't deppend on substitution parsing methods for testing the interpolator.
+# Write some dummy substitutions and parsing methods instead.
+
+
+def test_substitution_with_multiple_arguments():
+    subst = default_parse_substitution('$(env what heck)')
+    assert len(subst) == 1
+    subst = subst[0]
+    assert subst.name[0].perform(None) == 'what'
+    assert subst.default_value[0].perform(None) == 'heck'
+
+
+def test_escaped_characters():
+    subst = default_parse_substitution(r'$(env what/\$\(test asd\\\)) 10 10)')
+    assert len(subst) == 2
+    assert subst[0].name[0].perform(None) == 'what/$(test'
+    assert subst[0].default_value[0].perform(None) == r'asd\)'
+    assert subst[1].perform(None) == ' 10 10)'
+
+
+def test_nested_substitutions():
+    subst = default_parse_substitution('$(env what/$(test asd) 10) 10 10)')
+    assert len(subst) == 2
+    assert len(subst[0].name) == 2
+    assert subst[0].name[0].perform(None) == 'what/'
+    assert subst[0].name[1].perform(None) == 'asd'
+    assert subst[0].default_value[0].perform(None) == '10'
+    assert subst[1].perform(None) == ' 10 10)'
+
+
+def test_quoted_nested_substitution():
+    subst = default_parse_substitution(
+        'go_to_$(env WHERE asd)_of_$(env '
+        "'something $(test 10)')"
+    )
+    assert len(subst) == 4
+    assert subst[0].perform(None) == 'go_to_'
+    assert subst[1].name[0].perform(None) == 'WHERE'
+    assert subst[1].default_value[0].perform(None) == 'asd'
+    assert subst[2].perform(None) == '_of_'
+    assert subst[3].name[0].perform(None) == 'something '
+    assert subst[3].name[1].perform(None) == '10'
+    assert subst[3].default_value[0].perform(None) == ''
+
+
+def test_double_quoted_nested_substitution():
+    subst = default_parse_substitution(
+        r'$(env "asd_bsd_qsd_$(test \"asd_bds\")" "$(env DEFAULT)_qsd")'
+    )
+    context = LaunchContext()
+    assert len(subst) == 1
+    assert len(subst[0].name) == 2
+    assert subst[0].name[0].perform(context) == 'asd_bsd_qsd_'
+    assert subst[0].name[1].perform(context) == '"asd_bds"'
+    assert len(subst[0].default_value) == 2
+    assert subst[0].default_value[0].name[0].perform(context) == 'DEFAULT'
+    assert subst[0].default_value[0].default_value[0].perform(context) == ''
+    assert subst[0].default_value[1].perform(context) == '_qsd'
+
+
+def test_combining_quotes_nested_substitution():
+    subst = default_parse_substitution(
+        '$(env "asd_bsd_qsd_$(test \'asd_bds\')" \'$(env DEFAULT)_qsd\')'
+    )
```

Comment:
@hidmic I added this new test case. For having a more uniform behavior, I think this should look like:
`'$(env "asd_bsd_qsd_$(test \\\'asd_bds\\\')" \'$(env DEFAULT)_qsd\')'`
(this is not a raw a string, I'm escaping both the \ and ')

For that, I would change the following in the grammar:
```
SINGLE_QUOTED_STRING: (/[^'\\"$]|\$(?=!\()|(?<=\\)\$/ | "\\\"" | "\\'" | "\\")+
SINGLE_QUOTED_RSTRING: (/[^ '\\"$\(\)]|(?<=\\)\(|(?<=\\)\)|\$(?=!\()|(?<=\\)\$/ | "\\\"" | "\\'" | "\\")+

DOUBLE_QUOTED_STRING: (/[^"\\'$]|\$(?=!\()|(?<=\\)\$/ | "\\\"" | "\\'" | "\\")+
DOUBLE_QUOTED_RSTRING: (/[^ "\\'$\(\)]|(?<=\\)\(|(?<=\\)\)|\$(?=!\()|(?<=\\)\$/ | "\\\"" | "\\'" | "\\")+
```

Basically, I'm stopping allowing any kind of quote inside `SINGLE_QUOTED_*` and `DOUBLE_QUOTED_*`, and only allowing it if it is escaped.

Do you agree with that change?

Reply:
I'm fine with it, though it may be counter intuitive for some accustomed to combine quotes. If we do it, we should clearly document it somewhere.  @wjwwood any thoughts on this i.e. on forcing quote escaping everywhere if one wants to keep it raw?